contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenSource.sol

```diff
@@ -0,0 +1,207 @@
+            message,
+            (SourceAction, bytes)
+        );
+
+        // Route to the appropriate function.
+        if (action == SourceAction.Unlock) {
+            (address recipient, uint256 amount) = abi.decode(
+                actionData,
+                (address, uint256)
+            );
+            _unlockTokens(recipient, amount);
+        } else if (action == SourceAction.Burn) {
+            uint256 newBurnTotal = abi.decode(actionData, (uint256));
+            _handleBurnTokens(newBurnTotal);
+        } else {
+            revert("NativeTokenSource: invalid action");
+        }
+    }
+
+    /**
+     * @dev See {INativeTokenSource-transferToDestination}.
+     */
+    function transferToDestination(
+        address recipient,
+        TeleporterFeeInfo calldata feeInfo,
+        address[] calldata allowedRelayerAddresses
+    ) external payable nonReentrant {
+        // The recipient cannot be the zero address.
+        require(
+            recipient != address(0),
+            "NativeTokenSource: zero recipient address"
+        );
+
+        // Lock tokens in this bridge instance. Supports "fee/burn on transfer" ERC20 token
+        // implementations by only bridging the actual balance increase reflected by the call
+        // to transferFrom.
+        uint256 adjustedFeeAmount;
+        if (feeInfo.amount > 0) {
+            adjustedFeeAmount = SafeERC20TransferFrom.safeTransferFrom(
+                IERC20(feeInfo.feeTokenAddress),
+                feeInfo.amount
+            );
+            SafeERC20.safeIncreaseAllowance(
+                IERC20(feeInfo.feeTokenAddress),
+                address(teleporterMessenger),
+                adjustedFeeAmount
+            );
+        }
+
+        uint256 messageID = teleporterMessenger.sendCrossChainMessage(
+            TeleporterMessageInput({
+                destinationBlockchainID: destinationBlockchainID,
+                destinationAddress: nativeTokenDestinationAddress,
+                feeInfo: feeInfo,
+                requiredGasLimit: MINT_NATIVE_TOKENS_REQUIRED_GAS,
+                allowedRelayerAddresses: allowedRelayerAddresses,
+                message: abi.encode(recipient, msg.value)
+            })
+        );
+
+        emit TransferToDestination({
+            sender: msg.sender,
+            recipient: recipient,
+            amount: msg.value,
+            teleporterMessageID: messageID
+        });
+    }
+
+    /**
+     * @dev Unlocks tokens to recipient.
+     */
+    function _unlockTokens(address recipient, uint256 amount) private {
+        require(
+            recipient != address(0),
+            "NativeTokenSource: zero recipient address"
+        );
+        require(
+            address(this).balance >= amount,
+            "NativeTokenSource: insufficient collateral"
+        );
+
+        // Transfer to recipient
+        emit UnlockTokens(recipient, amount);
+        payable(recipient).transfer(amount);
```

Comment:
The OZ Defender report suggests using `recipient.call{value: amount, gas: XXX}("")` here. I think this is probably a worthwhile change because it's possible for the `fallback` method of a contract to be called to execute code. We have the proper `nonReentrant` guard in place already to prevent against any reentrancy, so I think we can use the full `gasleft()` amount in the call. 

This could also apply to the two other places we use `.transfer()` in the new contracts, and I'd say is worth updating these to be consistent, but given that they only are transferring to the burn address it shouldn't be possible for that to call code either way.

![image](https://github.com/ava-labs/teleporter/assets/55204436/9b572867-201e-42ce-83ad-5fc7dbd09898)

Reply:
@michaelkaplan13 What do you think of using https://github.com/OpenZeppelin/openzeppelin-contracts/blob/6ba452dea4258afe77726293435f10baf2bed265/contracts/utils/Address.sol#L41? 

Just using `recipient.call{value: amount, gas: XXX}("")` actually bloats to 
`       
(bool success, ) = BURNED_TX_FEES_ADDRESS.call{value: amount, gas: gasleft()}("");
 require(success, "NativeTokensource: failed to burn tokens");
`
Because we have to check the return status of `call`.

Reply:
I'm not 100% sure, but I think `.call()` passes along the remaining gas if not explicitly set also.

Reply:
Confirmed that [here](https://ethereum-blockchain-developer.com/2022-04-smart-wallet/07-low-level-calls-in-depth/).

```
You can also leave the gas amount, then it will forward all gas and let the called contract execute its logic and return the remainder. Safer is to provide an upper limit, just in case.
```

Great find with that OZ library, lets definitely use it 👍

Reply:
@michaelkaplan13 can you clarify your consideration of `fallback` functions in terms of comparing `tranfser` and `call`? IIUC, both transfer and low level call handle fallback functions the same, and both are guarded through reentrancy guards, so the difference  is the gas passed down

Reply:
My understanding is that `transfer` always uses at most 2300 gas, which creates a hard limit on the amount of gas available to any `fallback` function. That becomes problematic if the gas cost of operations change in network upgrades because fallback functions that may have used <=2300 gas prior may need >2300 afterwards. Using `call` lets them use as much gas is provided in the transaction, so they could still work even if they use more gas after future upgrades.